File: LiveDataExample/LiveDataExampleJava/app/src/main/java/com/tinlone/livedataexamplejava/PriceManager.java
package com.tinlone.livedataexamplejava;

import java.util.ArrayList;

public final class PriceManager {
    ArrayList<PriceListener> mListeners = new ArrayList<>();
    int price;

    private PriceManager() {
    }

    private static final class Holder {
        public static final PriceManager INSTANCE = new PriceManager();
    }

    public static PriceManager getInstance() {
        return Holder.INSTANCE;
    }

    public void requestPriceUpdate(PriceListener listener) {
        mListeners.add(listener);
        listener.onPriceChanged(price);
    }

    public void removeUpdateListener(PriceListener listener) {
        mListeners.remove(listener);
    }

    public void updatePrice(int price) {
        this.price = price;
        dispatchPriceChanged();
    }

    private void dispatchPriceChanged() {
        for (PriceListener listener : mListeners) {
            listener.onPriceChanged(price);
        }
    }

    public interface PriceListener {
        void onPriceChanged(int price);
    }
}

File: LiveDataExample/LiveDataExampleKotlin/app/src/main/java/com/tinlone/livedataexamplekotlin/PriceManager.kt
package com.tinlone.livedataexamplekotlin

import java.util.*

class PriceManager private constructor() {
    var mListeners = ArrayList<PriceListener>()
    var price = 0

    private object Holder {
        val instance = PriceManager()
    }

    fun requestPriceUpdate(listener: PriceListener) {
        mListeners.add(listener)
        listener.onPriceChanged(price)
    }

    fun removeUpdateListener(listener: PriceListener) {
        mListeners.remove(listener)
    }

    fun updatePrice(price: Int) {
        this.price = price
        dispatchPriceChanged()
    }

    private fun dispatchPriceChanged() {
        for (listener in mListeners) {
            listener.onPriceChanged(price)
        }
    }


    companion object {

        fun getInstance(): PriceManager {
           return Holder.instance
        }
    }
}